polar-core/src/bindings.rs

```diff
@@ -243,28 +259,35 @@ impl BindingManager {
         let mut op = op!(And, term.clone());
         for var in op.variables().iter().rev() {
             match self._variable_state(var) {
-                BindingManagerVariableState::Unbound => {}
                 BindingManagerVariableState::Cycle(c) => {
-                    let mut cycle = cycle_constraints(c);
-                    cycle.merge_constraints(op.clone());
-                    op = cycle;
+                    op = cycle_constraints(c).merge_constraints(op)
                 }
-                BindingManagerVariableState::Partial(e) => {
-                    let mut e = e.clone();
-                    e.merge_constraints(op);
-                    op = e;
-                }
-                BindingManagerVariableState::Bound(v) => {
-                    panic!(
-                        "Unexpected bound variable {var} in constraint. {var} = {val}",
-                        var = var,
-                        val = v
-                    );
+                BindingManagerVariableState::Partial(e) => op = e.clone().merge_constraints(op),
+                _ => {}
+            }
+        }
+
+        let vars = op.variables();
```

Comment:
Why is this a separate loop from the one above?

Reply:
Because the first loop may have added new variables to the expression.